src/scenic/syntax/veneer.py

```diff
@@ -115,996 +134,1320 @@
+        namespace.clear()
+        namespace.update(sampledNS)
 
 def endSimulation(sim):
-	global currentSimulation, currentScenario, currentBehavior, runningScenarios
-	global _globalParameters
-	currentSimulation = None
-	currentScenario = None
-	runningScenarios = set()
-	currentBehavior = None
-	_globalParameters = {}
-
-	for modName, (namespace, sampledNS, originalNS) in sim.scene.behaviorNamespaces.items():
-		namespace.clear()
-		namespace.update(originalNS)
+    global currentSimulation, currentScenario, currentBehavior, runningScenarios
+    global _globalParameters
+    currentSimulation = None
+    currentScenario = None
+    runningScenarios = set()
+    currentBehavior = None
+    _globalParameters = {}
+
+    for modName, (namespace, sampledNS, originalNS) in sim.scene.behaviorNamespaces.items():
+        namespace.clear()
+        namespace.update(originalNS)
 
 def simulationInProgress():
-	return currentSimulation is not None
+    return currentSimulation is not None
 
 # Requirements
 
 @contextmanager
-def executeInRequirement(scenario, boundEgo):
-	global evaluatingRequirement, currentScenario
-	assert activity == 0
-	assert not evaluatingRequirement
-	evaluatingRequirement = True
-	if currentScenario is None:
-		currentScenario = scenario
-		clearScenario = True
-	else:
-		assert currentScenario is scenario
-		clearScenario = False
-	oldEgo = currentScenario._ego
-	if boundEgo:
-		currentScenario._ego = boundEgo
-	try:
-		yield
-	except RandomControlFlowError as e:
-		# Such errors should not be possible inside a requirement, since all values
-		# should have already been sampled: something's gone wrong with our rebinding.
-		raise RuntimeError('internal error: requirement dependency not sampled') from e
-	finally:
-		evaluatingRequirement = False
-		currentScenario._ego = oldEgo
-		if clearScenario:
-			currentScenario = None
+def executeInRequirement(scenario, boundEgo, values):
+    global evaluatingRequirement, currentScenario
+    assert activity == 0
+    assert not evaluatingRequirement
+    evaluatingRequirement = True
+    if currentScenario is None:
+        currentScenario = scenario
+        clearScenario = True
+    else:
+        assert currentScenario is scenario
+        clearScenario = False
+    oldEgo = currentScenario._ego
+
+    currentScenario._sampledObjects = (values[obj] for obj in currentScenario.objects)
+
+    if boundEgo:
+        currentScenario._ego = boundEgo
+    try:
+        yield
+    except RandomControlFlowError as e:
+        # Such errors should not be possible inside a requirement, since all values
+        # should have already been sampled: something's gone wrong with our rebinding.
+        raise RuntimeError('internal error: requirement dependency not sampled') from e
+    finally:
+        evaluatingRequirement = False
+        currentScenario._ego = oldEgo
+        currentScenario._sampledObjects = currentScenario.objects
```

Comment:
I think it would be better to reset this to `None`, since that's what you use in `CanSee` to detect whether we have sampled values for the objects or not.